### ISSUE TITLE
Please consider fixing more bad APIs in v2.0

### ISSUE THREAD

**Issue by majewsky** [gophercloud/gophercloud#2902]
I saw that you put out a v2.0.0 beta, and the breaking changes are very tame. Since this is a unique opportunity, I wonder how many thoughts have been spent on improving the core APIs. Specifically, I have two requests:

- Please consider removing all the `Err4xxer` interfaces and `ErrDefault4xx` structs without replacement, and just returning `ErrUnexpectedResponseCode` instead. Or at the very least, these types should be documented much better to explain what their purpose is. As far as I can tell, all they do is [ruin good error messages unless you know how to get rid of them](https://github.com/sapcc/limes/blob/f25253aacf7849836af9a08d28278e4e8ae6dd60/internal/util/errors.go#L28-L36).

- The `context.Context` support added in the previous release is pretty much worthless since all the structured API methods do not take `ctx` arguments. This is the best possible moment to fix that, and add a `ctx` argument to every method that wraps `ProviderClient.Request` or `ServiceClient.Request`. The variants of those methods without `ctx` arguments should be removed.

This just off the top of my head, there's probably more. Please consider using this opportunity if at all possible.

**Comment by pierreprinetti** [comment]
Hi @majewsky, thank you for coming forward.

We have chosen to take a backwards-compatible approach to most changes, including the adoption of `context.Context`. Now that the main library is ready, we want to over time add `WithContext` versions of, ideally, every single call in Gophercloud. This is a time-consuming effort that you're welcome to join, and it can be done within `v2` (or even `v1`) without a major version bump.

The conversation on errors is a bit more complicated. I agree that Gophercloud errors need restructuring, and at the same time I'm afraid of breaking users in nasty ways. We didn't get complaints, so we assumed that we could just postpone the resolution of this issue. That's why I'm glad to hear from you. I would like all errors in Gophercloud to implement `Unwrap()` and expose strategic error types to facilitate their handling (in particular, but not limited to, the 404s). But I am afraid that we might be targeting v3 here.

Note that you're welcome to join [the #gophercloud channel in the Kubernetes Slack workspace](https://kubernetes.slack.com/archives/C05G4NJ6P6X/p1707471033444039).

**Comment by pierreprinetti** [comment]
@majewsky do you have a proposal for errors?

**Comment by majewsky** [comment]
> @majewsky do you have a proposal for errors?

As I said, just get rid of `Err4xxer` and `ErrDefault4xx` entirely and just return the plain `ErrUnexpectedResponseCode`. I don't see any advantage from those additional types. If you want to match on specific error codes and you find that

```
var uerr gophercloud.ErrUnexpectedResponseCode
if errors.As(err, &uerr) && uerr.Actual == 403 {
  handleForbidden()
}
```

is too tedious, you could do something like https://pkg.go.dev/github.com/majewsky/schwift#Is and provide a custom error matching function for matching a specific response code.

**Comment by majewsky** [comment]
I quickly wipped up a PR for the error-type thing (#2904) since that's something that I could do very quickly. Hopefully this either ends up simplifying the API, or improving clarity about why the larger API surface is actually necessary.

Re "adding `ctx` to everything", if I had a few days of time, I could see if I could do this is a tool-assisted mass change. This ought to be accessible to something slightly above regexes in power. What is your timeline for the 2.0 final release, just so I know?

**Comment by pierreprinetti** [comment]
>I could see if I could do this is a tool-assisted mass change

That would be ideal; writing a good tool is also contemplated within the "time-consuming effort" I mentioned above.

There is no deadline for adding `WithContext` functions, as it will be a completely backwards-compatible change. It can be included in `v2.0.0`, or in any `v2.y.0`.

**Comment by majewsky** [comment]
Since both #2904 and #2936 are merged, the scope from the original description is done.
